frontend/src/App.js

```diff
@@ -1,27 +1,51 @@
 import React, { useState, useEffect } from 'react';
-import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
+import {
+  BrowserRouter as Router,
+  Route,
+  Switch,
+  useLocation,
+  useHistory,
+} from 'react-router-dom';
+import jwt_decode from 'jwt-decode';
 import { ThemeProvider } from '@material-ui/core/styles';
 import CssBaseline from '@material-ui/core/CssBaseline';
+import Container from '@material-ui/core/Container';
+import Box from '@material-ui/core/Box';
 import theme from './styles/theme';
-import jwt_decode from 'jwt-decode';
 
 import Hypertube from './components/gallery/Index';
 import Landing from './components/layout/Landing';
 import Navbar from './components/layout/Navbar';
 import setAuthToken from './utils/setAuthToken';
-
-import Box from '@material-ui/core/Box';
-import { Container } from 'react-bootstrap';
+import authService from './services/auth';
 
 function App() {
   const [user, setUser] = useState({ userId: '', lang: '' });
+  const location = useLocation();
+  const history = useHistory();
 
   useEffect(() => {
-    setAuthToken(localStorage.getItem('token'));
-    if (localStorage.getItem('token')) {
-      const decoded = jwt_decode(localStorage.getItem('token'));
+    const setUserFromToken = (token) => {
+      const decoded = jwt_decode(token);
       setUser({ userId: decoded.id, lang: decoded.lang });
-    }
+			setAuthToken(token);
+    };
+
+    const checkToken = async () => {
+      try {
+        const { token } = await authService.getToken();
+        setUserFromToken(token);
+				history.push('/');
+      } catch (e) {
+        console.log(e);
+      }
+    };
+
+    if (location.search === '?auth=token') {
+      checkToken();
+    } else if (localStorage.getItem('token')) {
+      setUserFromToken(localStorage.getItem('token'));
```

Comment:
The token is available at the /token route only after signing in with google/42, so regular users would get null.
But you are right, it could be possible to get the token from localstorage for all user types. I was struggling with the sync/async inside the useEffect when I did this, but now that I think about it I could take another go tomorrow morning and put everything inside the async function!